Add surround keybinds

sudormrfbin opened the pull request:
Adds [vim-surround](https://github.com/tpope/vim-surround) functionality, with keybinds inspired from [vim-sandwich](https://github.com/machakann/vim-sandwich):

- ~`ma<char>`~ `ms<char>` - Add `char` around selection
- `md<char>` - Delete `char` pair
- `mr<from><to>` - Replace `from` with `to`

![helix-surround-2](https://user-images.githubusercontent.com/23398472/122865801-97073180-d344-11eb-8142-8f43809982c6.gif)

Goto matching bracket (previously `m`) has been moved to `mm`.

### TODO:

- [x] Add documentation
    - [x] Usage section
    - [x] Keymaps
    - [x] `mm` change
- [x] Add tests
- [x] [bug] Correctly identify pairs when cursor is _on_ a pair character

### Future work:

- `vim-sandwich` like visual feedback
- `b` as a generic surround char finder (like `b` in vim-sandwich)
- Show error when replace or delete cannot be done
- Treesitter based implementation ?

### Bugs

There are some subtle bugs right now with finding the correct surround pairs when they are spread across multiline lines, which can be solved by either using indent heuristics (works for most files) or directly querying treesitter, which is easier and accurate for languages with defined grammars (see `match_bracket` implementation).

pickfire commented:
Just wondering, how do you get the text showing there? Looks cool.

pickfire reviewed (commented):
Looks interesting but the keys looks weird. At least the key proposed here seemed inconsistent or does not work well with the current select option. What if we do

- select around tag (include the key searched)
- select inner tag (exclude the key searched)
- select only tag (only the 2 matched keys found)

Which only does select, and to change stuff we can do `d` or `c` or `i` after select. At least with these it is more consistent and can work with existing cursors or selection.

sudormrfbin commented:
> Just wondering, how do you get the text of what you typed showing there? Looks cool.

It's [screenkey](https://gitlab.com/screenkey/screenkey) :D

> Which only does select, and to change stuff we can do d or c or i after select. At least with these it is more consistent and can work with existing cursors or selection.

(I'm assuming when you say tags in the bullet points you mean surround characters like `(`, `[`, etc)

That's how I initially tried to implement it - placing selections on the surround chars seems the ideal way to do it, but I quickly ran into problems. Since the cursors placed on them are normal selections, deletion will work well, but adding and replacing will become hard.

If we have a buffer like this and the `|` are the cursor positions...

```
some |text|
```

... if the user presses `i(` we can reliably change `text` to `(text)`.

But if we have something like this...

```
|
|
```

... and the user tries to type `(text)` on each line, this is what they'll get on typing the first `(`:

```
(|
)|
```

The problem is we can't magically tell if the intention is to simply add a single `(` or surround with `()`. We'd need to implement some kind of mode for this, and then add hooks for insert, replace and change and add a way to select the surrounds.

I agree that selections you can make with respect to surround chars which you can then manipulate with normal operations like `i`, `c`, `d`, `r`, etc, are more intuitive, but right now they require too much work (unless there's a way to achieve it that I'm missing that doesn't involve too many hacks).

sudormrfbin commented:
> * select around tag (include the key searched)
> 
> * select inner tag (exclude the key searched)

Text objects for these two would be nice to have though.


> But one issue with this model that I mention is it won't work well on html end tags since they require an extra `/`.

HTML tag support is currently not implemented but definitely worth looking into in the future (the current implementation assumes all surround chars have a length of 1)

archseer reviewed (commented):
I'm trying to think through the mappings but I'd probably like to use `ma` and `mi` for text object selection. It feels like they'll be more frequently used too.

Maybe we could consider `m+`/`m-`? It's a bit harder to type though.. Or `ms`/`mr` for "match surround"/"match remove"

After thinking through usability and looking at https://github.com/h-youhei/kakoune-surround I do think you're right in surrounding the selection but not requiring a selection on remove, it seems easier to use because you don't need to make a precise selection.

pickfire commented:
Ah, I didn't know kakoune have something similar.

But one thing I am worried is what if someone wants to add/replace with multiple characters rather than one character? Do we not support surround cases like that? I see it being a good usecase like.

```html
  <p>one</p>
  <p>two</p>
  <p>three</p>
```

For example, adding `<strong>` and `</strong>` between wrapping around `one`, `two` and `three` and within `<p>`. I find myself needing to do this quite frequently or something similar at least in html. Maybe we have an uppercase version to add multiple characters?

pickfire commented on helix-core/src/surround.rs:
```diff
@@ -0,0 +1,58 @@
+use crate::{search, Selection};
+use ropey::RopeSlice;
+
+pub const PAIRS: &[(char, char)] = &[('(', ')'), ('[', ']'), ('{', '}'), ('<', '>')];
```
Let's add a few more which keyboard can usually type.
```suggestion
pub const PAIRS: &[(char, char)] = &[('(', ')'), ('[', ']'), ('{', '}'), ('<', '>'), ('«', '»'), ('「', '」), ('（', '）'), ('｛', '｝')];
```

sudormrfbin replied:
I've added them in. We should let this be extensible via the plugin system when it gets stabilized.

pickfire commented on helix-core/src/surround.rs:
```diff
@@ -0,0 +1,58 @@
+    for range in selection {
+        let head = range.head;
+
+        match find_nth_pairs_pos(text, ch, head, skip) {
```
I think you can just use `?` here to propagate `None`.

sudormrfbin replied:
Ah you're right !

pickfire reviewed (approved):
Looks good to me. I tested and it works with usual case, and I try breaking it with utf-8 characters but it still work. The only case where it didn't work is.
```html
<p *ngIf="hello < 1">w(#)orld</p>
<p *ngIf="hello < 1">w(#)orld</p>
<p *ngIf="hello < 1">w(#)orld</p>
```
With cursor positioned at `(#)`, when I do `mr<a`, it became
```html
<p *ngIf="hello a 1">world</pa
<p *ngIf="hello a 1">world</pa
<p *ngIf="hello a 1">world</pa
```
Yeah, we don't support this now, but I think this may do weird cases where people have quotes within these special characters and it will break some stuff on people's code if we are not careful, especially with parts that have regex.

pickfire commented:
Just wondering, can we have `mc<char>` to change a single character with multiple characters like what we have in normal mode?

archseer reviewed (changes requested):
Making sure this doesn't get merged accidentally, at the very least the `gm` conflict needs resolving

sudormrfbin commented on helix-term/src/commands.rs:
```diff
@@ -1758,6 +1759,7 @@ fn goto_mode(cx: &mut Context) {
             match (doc.mode, ch) {
                 (_, 'g') => move_file_start(cx),
                 (_, 'e') => move_file_end(cx),
+                (_, 'm') => match_brackets(cx),
```
`g%` seems too hard to type, how about `mm` or `mb` for "matching bracket" ?

archseer replied:
I think `mm` works for me because it's easy to just double tap. `mb` is indeed a good fit for "match brace" e.g. any closest matching brace (`(`/`[`, etc)

sudormrfbin replied:
I'll add `mm` then. As an aside, it'd be better if it could jump to the brace from anywhere inside of two braces as opposed to when only being _on_ the braces themselves.

archseer replied:
I was considering that as well but followed vim behavior first so it was predictable where the cursor jumped. I'd be open to change it though!

sudormrfbin commented:
> Maybe we could consider `m+`/`m-`? It's a bit harder to type though.. Or `ms`/`mr` for "match surround"/"match remove"

@archseer the + and - variants are definitely hard to type - `ms` seems good for adding surrounds.

> But one thing I am worried is what if someone wants to add/replace with multiple characters rather than one character? Do we not support surround cases like that? I see it being a good usecase like.

@pickfire It's definitely a good and valid usecase (and a godsend when you're editing html). I'll add it in another PR since I don't want to stall this one :P 

> We may also best support multiple selections to be able to modify the three directly.

That's already implemented in this PR, it's showcased towards the end of the gif :)

> With cursor positioned at `(#)`, when I do `mr<a`, it became
> 
> ```
> <p *ngIf="hello a 1">world</pa
> <p *ngIf="hello a 1">world</pa
> <p *ngIf="hello a 1">world</pa
> ```

Yeah that's definitely a bug. I'm thinking of a treesitter based implementation and then using  the current implementation as a fallback. That would correctly identify the valid surround pairs.

> Just wondering, can we have `mc<char>` to change a single character with multiple characters like what we have in normal mode?

Aha that seems the most ergonomic way to do it; I'll add it with the html tag editing support PR :)

CBenoit commented:
(Passing by just to say it's really cool :clap: )

archseer commented on helix-term/src/keymap.rs:
```diff
@@ -202,7 +202,7 @@ impl Default for Keymaps {
 
-            key!('m') => Command::match_brackets,
+            key!('m') => Command::surround,
             // TODO: refactor into
```
Can you kill these comments? They're outdated

sudormrfbin replied:
Done 👍🏾

archseer commented on helix-term/src/commands.rs:
```diff
@@ -3061,6 +3062,123 @@ fn right_bracket_mode(cx: &mut Context) {
+
+            let mut changes = Vec::new();
+            for (i, range) in selection.iter().enumerate() {
+                let (from, to) = (range.from(), range.to() + 1);
```
Need to clamp the range to `max_to` here, like elsewhere:

```
let max_to = doc.text().len_chars().saturating_sub(1);
let to = std::cmp::min(range.to() + 1, max_to)
```

This is a bit different now on master with the CRLF fixes, you might want to do `get_line_ending(text).len_chars()`

sudormrfbin replied:
Clamped with `get_line_ending` 👍🏾

sudormrfbin commented on helix-term/src/keymap.rs:
```diff
@@ -200,17 +200,7 @@ impl Default for Keymaps {
-
-            // TODO: figure out what key to use
-            // key!('[') => Command::expand_selection, ??
+            key!('m') => Command::surround,
```
`mm` is goto matching bracket and `ma` and `mi` might be added for around and inside objects, so we should call this something else. `match_mode` maybe ?

archseer replied:
`match_mode` sounds good. Still need to remap `ma` -> `ms` then we're good 👍🏻

archseer replied:
(The expand selection key that's commented out above should probably be `[o`/`]o` for expand/shrink. Shrink is unimplemented)

sudormrfbin replied:
Done

sudormrfbin commented:
Added [documentation](https://github.com/helix-editor/helix/pull/320/files#diff-492d9e0201bf57b23dac04280ceff4f15735ac3781da61f9d0c43bc40f5d980c), changed `ma` to `ms` and rebased 🎉

archseer reviewed (approved):
🎉
